### ISSUE TITLE
Skipping tests in test spec

### ISSUE THREAD

**Issue by sirihansen** [ferd/cth_readable#10]
This is mostly for information - at least for now. The main problem is in `common_test`, but when that is corrected it might require an update of `cth_readable_shell`.

Since the hook function `on_tc_skip/3` does not include the name of the suite, `cth_readable_shell` is depending on `pre_init_per_suite/3` to get the suite name and store it in the hook state. The problem is that when a suite is skipped by specifying a `skip_suites` statement in the test spec, there is no call to `pre_init_per_suite/3` or any other hook function except `on_tc_skip/3` (with `TC=all`). The consequence of this is that `cth_readable_shell` will print skip info for either the previous suite (if there is one), or for the suite `undefined` (if the skipped suite is the first to be handled):

```
%%% undefined ==> all: SKIPPED
%%% undefined ==> {tc_user_skip,"...skip reason..."}
```
There is currently no workaround for this problem, but it might be a good idea to 'unset' the suite name in the state on `post_end_per_suite/4` in order to avoid the faulty skip info if a previous suite existed, for example:

```
%%% reltool_app_SUITE ==> fields: OK
%%% reltool_app_SUITE ==> modules: OK
%%% reltool_app_SUITE ==> export_all: OK
%%% reltool_app_SUITE ==> app_depend: OK
%%% reltool_app_SUITE ==> undef_funcs: OK
%%% reltool_app_SUITE ==> appup: OK
%%% reltool_app_SUITE ==> all: SKIPPED                            <-----------
%%% reltool_app_SUITE ==> {tc_user_skip,"...skip reason..."}      <-----------
```

**Comment by ferd** [comment]
Yikes, that's a good catch. I'm guessing that yeah, the best we can do there is what you mentioned and unset the suite in `post_end_per_suite` to avoid issues of misreporting. Better to lack info than flat out output wrong one.

The other thing I'd really like to be able to do is get group information, but the group mechanisms in CT are nestable (stackable?) to large extents where I just decided to omit the information for now.

**Comment by sirihansen** [comment]
A fix in common_test is planned for the next sprint and will be included in 19.3. Haven't started the job for real, but the alternative ways that we have discussed so far are

1. make sure pre_init_per_suite is always called. This might be a rather big change to core functionality in test_server, so not really risk free... There is also the question of whether it is really correct to call this hook when Suite:init_per_suite will not be called (which it won't, since the suite is skipped completely)

2. extend on_tc_skip to include the suite name as a separate argument, and fall back to the old function if the new is not exported.

3. add a new hook - on_suite_skip - and keep on_tc_skip unchanged. Might have to call both on_suite_skip and on_tc_skip (with TC=all) when suite is skipped, for backwards compatibility.

For now I'm leaning towards alternative 2, but might of course change my mind as we start the job for real. Any thoughts are welcome!

And yes, the group mechanism is a bit too flexible to be easily visualized... :(

**Comment by sirihansen** [comment]
FYI: The solution we chose was alternative 2 above - i.e. a `Suite` argument is added to `on_tc_skip` (and a lot of other hook callbacks). Fallback to old API if the new function is not exported. This is included in OTP-19.3, common_test-1.14.

**Comment by ferd** [comment]
Cool. I'll reopen this to track it and add support for 19.3
